src/huggingface_hub/inference/_mcp/mcp_client.py

```diff
@@ -0,0 +1,236 @@
+        """
+        from mcp import ClientSession, StdioServerParameters
+        from mcp import types as mcp_types
+        from mcp.client.stdio import stdio_client
+
+        logger.info(f"Connecting to MCP server with command: {command} {args}")
+        server_params = StdioServerParameters(
+            command=command,
+            args=args if args is not None else [],
+            env=env,
+            cwd=cwd,
+        )
+
+        stdio_transport = await self.exit_stack.enter_async_context(stdio_client(server_params))
+        stdio, write = stdio_transport
+        session = await self.exit_stack.enter_async_context(
+            ClientSession(
+                read_stream=stdio,
+                write_stream=write,
+                client_info=mcp_types.Implementation(
+                    name="huggingface_hub.MCPClient",
+                    version=get_hf_hub_version(),
+                ),
+            )
+        )
+
+        logger.debug("Initializing session...")
+        await session.initialize()
+
+        # List available tools
+        response = await session.list_tools()
+        logger.debug("Connected to server with tools:", [tool.name for tool in response.tools])
+
+        for tool in response.tools:
+            if tool.name in self.sessions:
+                logger.warning(f"Tool '{tool.name}' already defined by another server. Skipping.")
+                continue
+
+            # Map tool names to their server for later lookup
+            self.sessions[tool.name] = session
+
+            # Add tool to the list of available tools (for use in chat completions)
+            self.available_tools.append(
+                ChatCompletionInputTool.parse_obj_as_instance(
+                    {
+                        "type": "function",
+                        "function": {
+                            "name": tool.name,
+                            "description": tool.description,
+                            "parameters": tool.inputSchema,
+                        },
+                    }
+                )
+            )
+
+    async def process_single_turn_with_tools(
+        self,
+        messages: List[Union[Dict, ChatCompletionInputMessage]],
+        exit_loop_tools: Optional[List[ChatCompletionInputTool]] = None,
+        exit_if_first_chunk_no_tool: bool = False,
+    ) -> AsyncIterable[Union[ChatCompletionStreamOutput, ChatCompletionInputMessage]]:
+        """Process a query using `self.model` and available tools, yielding chunks and tool outputs.
+
+        Args:
+            query: The user query to process
+            opts: Optional parameters including:
+                - exit_loop_tools: List of tools that should exit the generator when called
+                - exit_if_first_chunk_no_tool: Exit if no tool is present in the first chunks
+
+        Yields:
+            ChatCompletionStreamOutput chunks or ChatCompletionInputMessage objects
+        """
```

Comment:
This docstring needs to be updated: `self.model` should be replaced by `self.client` and the args should be updated as well.